docs: add new routing in-depth guides

bencodezen opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?

There are now updated Routing guides!


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

JeanMeche commented on adev/src/content/guide/routing/read-route-state.md:
````diff
@@ -0,0 +1,280 @@
+}
+```
+
+For more information, check out the official docs for [isActiveMatchOptions](https://angular.dev/api/router/IsActiveMatchOptions).
````
```suggestion
For more information, check out the official docs for [isActiveMatchOptions](/api/router/IsActiveMatchOptions).
```

JeanMeche commented on adev/src/content/guide/routing/show-routes-with-outlets.md:
````diff
@@ -0,0 +1,168 @@
+You can add event listeners with the standard event binding syntax:
+
+```angular-html
+<router-outlet
+  (activate)='onActivate($event)'
+  (deactivate)='onDeactivate($event)'
+  (attach)='onAttach($event)'
+  (detach)='onDetach($event)'
+></router-outlet>
````
Should we make it a self-closing element ?

JeanMeche commented on adev/src/content/guide/routing/show-routes-with-outlets.md:
```diff
@@ -0,0 +1,168 @@
+
+## Next steps
+
+Check out the [API docs for RouterOutlet](https://angular.dev/api/router/RouterOutlet?tab=api) if you’d like to learn more.
```
```suggestion
Check out the [API docs for RouterOutlet](/api/router/RouterOutlet?tab=api) if you’d like to learn more.
```

JeanMeche commented on adev/src/content/guide/routing/read-route-state.md:
````diff
@@ -0,0 +1,280 @@
+</div>
+```
+
+For more information, check out the [API docs for RouterLinkActive](https://angular.dev/api/router/RouterLinkActive).
````
```suggestion
For more information, check out the [API docs for RouterLinkActive](/api/router/RouterLinkActive).
```

JeanMeche commented on adev/src/content/guide/routing/define-routes.md:
```diff
@@ -0,0 +1,401 @@
+
+If you modify or remove a route, some users may still click on out-of-date links or bookmarks to that route. You can add a redirect to direct those users to an appropriate alternative route instead of a "not found" page.
+
+<!-- TODO: Add link for redirecting routes guide -->
```
remaining todo

bencodezen commented on adev/src/content/guide/routing/define-routes.md:
````diff
@@ -0,0 +1,395 @@
+
+Nested routes, also known as child routes, are a common technique for managing more complex navigation routes where a component has a sub-view that changes based on the URL.
+
+```
````
TODO: Replace ASCII diagram with image to address a11y issues and readability.

michael-small commented on adev/src/content/guide/routing/define-routes.md:
```diff
@@ -0,0 +1,395 @@
+
+Most projects define routes in a separate file that contains `routes` in the filename.
+
+A collection of routes looks like this::
```
Extra `:`

bencodezen replied:
Thanks @michael-small!

JeanMeche commented on adev/src/content/guide/routing/read-route-state.md:
```diff
@@ -0,0 +1,295 @@
+export class ProductComponent {
+  private activatedRoute = inject(ActivatedRoute);
+
+  ngOnInit() {
```
We don't need to rely on the onInit hook. Current recommentation is to use `constructor` more. 
```suggestion
  constructor() {
```

JeanMeche commented on adev/src/content/guide/routing/define-routes.md:
```diff
@@ -0,0 +1,395 @@
+  // The HomePage and LoginPage components are loaded lazily at the point at which
+  // their corresponding routes become active.
+  {path: "login", loadComponent: () => import("./components/auth/login-page")},
+  {path: "", loadComponent: () => import("./components/home/home-page")}
```
Should we add the `then(m => m.HomePageCompoent)`?  
I don't think the default export of the component is part of our recommendations.

bencodezen replied:
@JeanMeche I'm not sure what you're referring to. Is there an additional render step to the returned component?

JeanMeche replied:
It should be 

`{path: "", loadComponent: () => import("./components/home/home-page").then(m => m.HomePage))`

bencodezen replied:
Ah got it. Appreciate the clarification!

JeanMeche commented on adev/src/content/guide/routing/read-route-state.md:
```diff
@@ -0,0 +1,295 @@
+| `url`           | An `Observable` of the route paths, represented as an array of strings for each part of the route path.                                                                                                   |
+| `data`          | An `Observable` that contains the `data` object provided for the route. Also contains any resolved values from the resolve guard.                                                                         |
+| `params`        | An `Observable` that contains the required and optional parameters specific to the route.                                                                                                                 |
+| `paramMap`      | An `Observable` that contains a [map](api/router/ParamMap) of the required and optional parameters specific to the route. The map supports retrieving single and multiple values from the same parameter. |
```
```suggestion
| `paramMap`      | An `Observable` that contains a [map](/api/router/ParamMap) of the required and optional parameters specific to the route. The map supports retrieving single and multiple values from the same parameter. |
```

mauriziocescon commented on adev/src/content/guide/routing/define-routes.md:
````diff
@@ -0,0 +1,395 @@
+
+```angular-ts
+import { AppAdminComponent } from './app-admin/app-admin.component';
+
````
```suggestion
import { AdminPage } from './app-admin/app-admin.component';
```

mauriziocescon commented on adev/src/content/guide/routing/define-routes.md:
```diff
@@ -0,0 +1,395 @@
+
+## Associating data with routes
+
+Route data enables developers to attach additional information to routes. Developers are able to configure how components behave based on this data.
```
```suggestion
Route data enables you to attach additional information to routes. You are able to configure how components behave based on this data.
```

mauriziocescon commented on adev/src/content/guide/routing/define-routes.md:
```diff
@@ -0,0 +1,395 @@
+
+When you need to provide dynamic data to a route, check out the [guide on route data resolvers](/guide/router/route-data-resolvers).
+
+Should I use an eager or a lazy route?
```
Likely a left over...

bencodezen replied:
Thanks for catching this @mauriziocescon!

mauriziocescon commented on adev/src/content/guide/routing/navigate-to-routes.md:
```diff
@@ -0,0 +1,163 @@
+In this example, the first example contains the full path with the protocol (i.e., `https://`) and the root domain (i.e., `angular.dev`) explicitly defined for the essentials page. In contrast, the second example assumes the user is already on the correct root domain for `/essentials`.
+
+Generally speaking, relative URLs are preferred because they are more maintainable across applications because they don’t need to know their absolute position within the routing hierarchy.
+How relative URLs work
```
```suggestion

### How relative URLs work
```

bencodezen commented:
Big thanks for @JeanMeche, @michael-small and @mauriziocescon for your detailed review. Appreciate y'all catching things that got missed in the transfer from Google Docs over to Markdown!

bencodezen committed:
docs: add new routing in-depth guide

docs: update code snippet

Co-authored-by: Andrew Scott <atscott01@gmail.com>

docs: update link

Co-authored-by: Matthieu Riegler <kyro38@gmail.com>

docs: update syntax highlighting

Co-authored-by: Andrew Scott <atscott01@gmail.com>

docs: correct example links

docs: update load component with more concise syntax

docs: fix link text

docs: fix api docs link formatting

docs: fix callout box types and rendering

docs: truncate properties table

docs: swap constructor routes to use inject instead

docs: add other routing tasks info in subnav and remove duplicate info

docs: fix type conversion param

JeanMeche commented on adev/src/content/guide/routing/define-routes.md:
````diff
@@ -0,0 +1,399 @@
+];
+```
+
+In this example, URLs such as `/user/leeroy` and `/user/jenkins` render the `UserProfile` component. This component can then read the `id` parameter and use it to perform additional work, such as fetching data. See [Reading route state](/guide/router/reading-route-state) for details on reading route parameters.
````
```suggestion
In this example, URLs such as `/user/leeroy` and `/user/jenkins` render the `UserProfile` component. This component can then read the `id` parameter and use it to perform additional work, such as fetching data. See [Reading route state](/guide/routing/read-route-state) for details on reading route parameters.
```

JeanMeche reviewed (approved):
Beside my comments LGTM

atscott commented on adev/src/content/guide/routing/define-routes.md:
```diff
@@ -0,0 +1,399 @@
+
+In this routes array, the app displays the `NotFound` component when the user visits any path outside of `home` and `user/:id`.
+
+HELPFUL: Wildcard routes are typically placed at the end of a routes array.
```
I don't know if this is "helpful" and "typically" as much as it is important and required (unless you have a `canMatch` guard I guess) due to the "first match wins".  The example right below here pretty much says just that "// Wildcard - always last". Here's the note in the old ToH guide: https://v16.angular.io/guide/router-tutorial-toh#define-a-wildcard-route.

bencodezen replied:
I've updated it to a "Tip" in order to call it out since this is something that I've seen users mess up before when defining routes and calling it out can help raise awareness of that fact.

atscott commented on adev/src/content/guide/routing/define-routes.md:
```diff
@@ -0,0 +1,399 @@
+  // their corresponding routes become active.
+  {
+    path: "login",
+    loadComponent: () => import("./components/auth/login-page").then(m => m.LoginPage)
```
We added support for default exports to make this shorter. It might be nicer to show that instead? 

`loadComponent: () => import("./components/auth/login-page")`

atscott commented on adev/src/content/guide/routing/define-routes.md:
```diff
@@ -0,0 +1,399 @@
+
+In general, eager loading is recommended for primary landing page(s) while other pages would be lazy-loaded.
+
+WARNING: While lazy routes have the upfront performance benefit of reducing the amount of initial data requested by the user, it adds future data requests that could be undesirable. This is particularly true when dealing with nested lazy loading at multiple levels, which can significantly impact performance.
```
Looks like we don't have a WARNING card. https://github.com/angular/angular/blob/9b0c9ffdf63a1916d027342db2737e435da59460/adev/shared-docs/pipeline/guides/extensions/docs-alert.ts#L12-L22 Not sure if this should be one of those for formatting, if we should _add_ a warning card, or if you're okay with the current state

bencodezen replied:
Good catch! For some reason I was under the impression we had one, but I can definitely make do instead with the note instead.

atscott commented on adev/src/content/guide/routing/show-routes-with-outlets.md:
````diff
@@ -0,0 +1,170 @@
+
+When a user visits `/products`, Angular renders the following:
+
+```angular-html
+<app-header></app-header>
+<app-products></app-products>
+<app-footer></app-footer>
+```
+
+If the user goes back to the home page, then Angular renders:
+
+```angular-html
+<app-header></app-header>
+<app-home></app-home>
+<app-footer></app-footer>
+```
````
nit: should these use self closing tags for consistency and brevity?

bencodezen replied:
Happy to do so for the first one, but the rendered example would should the closing tag in the element inspector if I'm not mistaken.

bencodezen replied:
I normally would agree, but since the example is illustrating what renders in the element inspector, I believe it'll always show the closing tag. Let me know if I'm mistaken though!

atscott commented on adev/src/content/guide/routing/read-route-state.md:
```diff
@@ -0,0 +1,295 @@
+
+The `ActivatedRoute` can provide different information about the route including:
+
+| Property        | Details                                                                                                                                                                                                    |
+| :-------------- | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `url`           | An `Observable` of the route paths, represented as an array of strings for each part of the route path.                                                                                                    |
+| `data`          | An `Observable` that contains the `data` object provided for the route. Also contains any resolved values from the resolve guard.                                                                          |
+| `params`        | An `Observable` that contains the required and optional parameters specific to the route.                                                                                                                  |
+| `paramMap`      | An `Observable` that contains a [map](/api/router/ParamMap) of the required and optional parameters specific to the route. The map supports retrieving single and multiple values from the same parameter. |
+| `queryParamMap` | An `Observable` that contains a [map](api/router/ParamMap) of the query parameters available to all routes. The map supports retrieving single and multiple values from the query parameter.               |
+| `queryParams`   | An `Observable` that contains the query parameters available to all routes.                                                                                                                                |
+| `fragment`      | An `Observable` of the URL fragment available to all routes.                                                                                                                                               |
+| `outlet`        | The name of the `RouterOutlet` used to render the route. For an unnamed outlet, the outlet name is primary.                                                                                                |
+| `routeConfig`   | The route configuration used for the route that contains the origin path.                                                                                                                                  |
+| `parent`        | The route's parent `ActivatedRoute` when this route is a child route.                                                                                                                                      |
+| `firstChild`    | Contains the first `ActivatedRoute` in the list of this route's child routes.                                                                                                                              |
+| `children`      | Contains all the child routes activated under the current route.                                                                                                                                           |
```
Hmmm, is it useful to duplicate the API reference here? Seems like it's not much more than a copy of it and would mean needing to keep it up to date.

bencodezen replied:
Fair point! I'd like to expose some of the properties to raise some awareness, so I've truncated it to the common properties people often use which should make this more manageable. Let me know if you have another suggestion you prefer!

atscott commented on adev/src/content/guide/routing/read-route-state.md:
```diff
@@ -0,0 +1,295 @@
+export class UserProfileComponent {
+  userId: string;
+
+  constructor(private route: ActivatedRoute) {
```
Should this use `inject` instead?

bencodezen replied:
Changing all examples to use inject!

atscott commented on adev/src/content/guide/routing/read-route-state.md:
```diff
@@ -0,0 +1,295 @@
+export class ProductDetailComponent {
+  productId: string;
+
+  constructor(private route: ActivatedRoute) {
```
nit: use `inject`?

atscott commented on adev/src/content/guide/routing/read-route-state.md:
```diff
@@ -0,0 +1,295 @@
+  `
+})
+export class ProductDetailComponent {
+  productId: string;
```
Should we use a signal here? This example isn't compatible with `OnPush` or zoneless

bencodezen replied:
I was not aware of that! Let's definitely make that change.

atscott commented on adev/src/content/guide/routing/read-route-state.md:
```diff
@@ -0,0 +1,295 @@
+
+### Route matching strategy
+
+By default, RouterLinkActive considers any ancestors in the route a match.
```
```suggestion
By default, `RouterLinkActive` considers any ancestors in the route a match.
```
nit: formatting

atscott commented on adev/src/content/guide/routing/navigate-to-routes.md:
````diff
@@ -0,0 +1,164 @@
+**Relative URLs** in Angular routing allow you to define navigation paths relative to the current route's location. This is in contrast to **absolute URLs** which contain the full path with the protocol (e.g., `http://`) and the **root domain** (e.g., `google.com`).
+
+```angular-html
+<!-- Absolute URL -->
+<a href="/essentials">Angular Essentials Guide</a>
+
+<!-- Relative URL -->
+<a href="/essentials">Angular Essentials Guide</a>
````
These examples are the same

bencodezen replied:
🤦‍♂️ Good catch!

atscott commented on adev/src/content/guide/routing/read-route-state.md:
```diff
@@ -0,0 +1,295 @@
+
+@Component({ ... })
+export class UserProfileComponent {
+  userId: string;
```
```suggestion
  readonly userId: string;
```

bencodezen commented on adev/src/content/guide/routing/read-route-state.md:
````diff
@@ -0,0 +1,295 @@
+}
+```
+
+Check out the [`ActivatedRouteSnapshot API docs`](/api/router/ActivatedRouteSnapshot) for a complete list of all properties you can access.
````
Got it!

bencodezen commented:
@atscott Alright! I think I've addressed all your comments and updated it to better reflect the inject pattern with signals in the code examples. Let me know if there's anything else you want me to change!

atscott commented on adev/src/content/guide/routing/read-route-state.md:
```diff
@@ -0,0 +1,288 @@
+
+Check out the [`ActivatedRoute` API docs](/api/router/ActivatedRoute) and [`ActivatedRouteSnapshot` API docs](/api/router/ActivatedRouteSnapshot) for a complete list of all properties you can access.
+
+## Reading parameters on a route
```
Describing how to use `ActivatedRoute` is definitely useful. I wonder if we should put more/equal emphasis on the `withComponentInputBinding` feature, which simplifies the story quite a bit.

```
  productId = signal('');
  private activatedRoute = inject(ActivatedRoute);
  constructor() {
    // Access route parameters
    this.activatedRoute.params.subscribe((params) => {
      this.productId.set(params['id']);
    });
  }
```

would instead be 

```
  productId = input.required<string>({alias: 'id'})
```

or 

```
  id = input.required<string>()
```

bencodezen replied:
I definitely would like to explore this more! Let's table it for another PR when we do another update on the routing docs. Will DM you more about this!

atscott commented on adev/src/content/guide/routing/read-route-state.md:
```diff
@@ -0,0 +1,288 @@
+    // Access query parameters reactively
+    this.route.queryParams.subscribe(params => {
+      const sort = params['sort'] || 'price';
+      const page = params['page'] || 1;
```
I think this should be converting `params['page']` to a number with `Number(params['page'])`? The default serializer doesn't convert things to numbers IIRC. (Also page 0 wouldn't be reachable with this logic). Alternatively, the page could just be a string...

Following the previous comment about input binding, this could be 


```
  sort = input({transform: (value?: string) => value || 'price'});
  page = input({transform: (value?: string) => value !== undefined ? Number(value) : 1});
  constructor() {
    effect(() => {
      this.loadProducts(this.sort(), this.page());
    });
  }
```

bencodezen replied:
Huh... I definitely want to double tap more on this `input` function because I was always under the impression it was only for forms, so I'll bookmark this to chat more with you separately as well.

In the meantime, I'll add the type conversion function to keep the code consistent!

JeanMeche commented:
@bencodezen Could you squash your commits, drop the merge commits and rebase on main please. The PR is not mergeable as is.

pkozlowski-opensource commented:
This PR was merged into the repository by commit 3a6121fb866fad5792d409fd60879b9041d955b2.

The changes were merged into the following branches: main
